In [148]:
import pandas as pd
import matplotlib.pyplot as plt

In [149]:
mydata = pd.read_excel("/Users/amitmishra/Desktop/proj/liverpool 2022-23 defence.xlsx")

In [150]:
mydata.head(3)

Date           Time              Comp                Round  Day  \
0 2022-07-30  17:00 (11:00)  Community Shield  FA Community Shield  Sat   
1 2022-08-06  12:30 (06:30)    Premier League          Matchweek 1  Sat   
2 2022-08-15  20:00 (14:00)    Premier League          Matchweek 2  Mon   

     Venue Result GF GA         Opponent  ...  Int  Tkl+Int   Clr  Err  GCA  \
0  Neutral      W  3  1  Manchester City  ...    8        0   NaN  NaN  NaN   
1     Away      D  2  2           Fulham  ...   10       21  16.0  0.0  4.0   
2     Home      D  1  1   Crystal Palace  ...   13       33   9.0  0.0  2.0   

    SCA  TouchesDef  TouchesMid  TouchesAtt  TouchesPen  
0   NaN         NaN         NaN         NaN         NaN  
1  20.0       260.0       349.0       179.0        33.0  
2  45.0       147.0       527.0       244.0        42.0  

[3 rows x 32 columns]

In [151]:
mydata.isna().sum()

Date          0
Time          0
Comp          0
Round         0
Day           0
Venue         0
Result        0
GF            0
GA            0
Opponent      0
Tkl           6
TklW          0
Def 3rd       6
Mid 3rd       6
Att 3rd       6
Tkl.1         6
Att           6
Tkl%          6
Lost          6
Blocks        6
Sh            6
Pass          6
Int           0
Tkl+Int       0
Clr           6
Err           6
GCA           6
SCA           6
TouchesDef    6
TouchesMid    6
TouchesAtt    6
TouchesPen    6
dtype: int64

### In the absence of data from some competitions, let's drop the entire row with any null values

In [152]:
cleaned_data = mydata.dropna()

In [156]:
cleaned_data.isna().sum()

Date          0
Time          0
Comp          0
Round         0
Day           0
Venue         0
Result        0
GF            0
GA            0
Opponent      0
Tkl           0
TklW          0
Def 3rd       0
Mid 3rd       0
Att 3rd       0
Tkl.1         0
Att           0
Tkl%          0
Lost          0
Blocks        0
Sh            0
Pass          0
Int           0
Tkl+Int       0
Clr           0
Err           0
GCA           0
SCA           0
TouchesDef    0
TouchesMid    0
TouchesAtt    0
TouchesPen    0
dtype: int64

In [157]:
print(cleaned_data.columns.tolist())

['Date', 'Time', 'Comp', 'Round', 'Day', 'Venue', 'Result', 'GF', 'GA', 'Opponent', 'Tkl', 'TklW', 'Def 3rd', 'Mid 3rd', 'Att 3rd', 'Tkl.1', 'Att', 'Tkl%', 'Lost', 'Blocks', 'Sh', 'Pass', 'Int', 'Tkl+Int', 'Clr', 'Err', 'GCA', 'SCA', 'TouchesDef', 'TouchesMid', 'TouchesAtt', 'TouchesPen']


In [158]:
cleaned_data = mydata.rename(columns = {'Def 3rd':'TklDef', 'Mid 3rd':'TklMid', 'Att 3rd':'TklAtt','Sh':'ShotBlocked', 'Pass':'PassBlocked'})

In [159]:
cleaned_data.columns

Index(['Date', 'Time', 'Comp', 'Round', 'Day', 'Venue', 'Result', 'GF', 'GA',
       'Opponent', 'Tkl', 'TklW', 'TklDef', 'TklMid', 'TklAtt', 'Tkl.1', 'Att',
       'Tkl%', 'Lost', 'Blocks', 'ShotBlocked', 'PassBlocked', 'Int',
       'Tkl+Int', 'Clr', 'Err', 'GCA', 'SCA', 'TouchesDef', 'TouchesMid',
       'TouchesAtt', 'TouchesPen'],
      dtype='object')

In [93]:
mycols = ['GA','TklDef', 'TklMid', 'TklAtt','ShotBlocked', 'PassBlocked','Clr']

In [160]:
colsattack = ['GF','GCA', 'SCA', 'TouchesDef', 'TouchesMid','TouchesAtt', 'TouchesPen']

In [94]:
df = cleaned_data[mycols]

In [95]:
df.isna().sum()

GA             0
TklDef         0
TklMid         0
TklAtt         0
ShotBlocked    0
PassBlocked    0
Clr            0
dtype: int64

In [96]:
df = pd.DataFrame(df)

In [97]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [117]:
X = df[['TklDef', 'TklMid', 'TklAtt','ShotBlocked','Clr']]

In [118]:
y = df['GA']

In [100]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.3, random_state = 42)

In [101]:
model_reg = LinearRegression()

In [102]:
model_reg.fit(X_train,y_train)

LinearRegression()

In [103]:
y_pred = model_reg.predict(X_test)

In [108]:
import statsmodels.api as smf

In [123]:
df.corr()

GA    TklDef    TklMid    TklAtt  ShotBlocked  PassBlocked  \
GA           1.000000 -0.184648  0.109749  0.119957     0.025374    -0.070865   
TklDef      -0.184648  1.000000  0.011200 -0.295435     0.384433     0.242269   
TklMid       0.109749  0.011200  1.000000  0.035853    -0.087302     0.009313   
TklAtt       0.119957 -0.295435  0.035853  1.000000    -0.398513    -0.119743   
ShotBlocked  0.025374  0.384433 -0.087302 -0.398513     1.000000     0.110746   
PassBlocked -0.070865  0.242269  0.009313 -0.119743     0.110746     1.000000   
Clr         -0.197161  0.284121 -0.189122 -0.281631     0.405043    -0.048640   

                  Clr  
GA          -0.197161  
TklDef       0.284121  
TklMid      -0.189122  
TklAtt      -0.281631  
ShotBlocked  0.405043  
PassBlocked -0.048640  
Clr          1.000000

In [113]:
# Assuming df is your DataFrame and 'column_name' is the column you want to convert
df[mycols] = df[mycols].astype(int)

In [119]:
X = smf.add_constant(X)

In [120]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.3, random_state = 42)

In [121]:
model = smf.OLS(y_train,X_train).fit()

In [124]:
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                     GA   R-squared:                       0.233
Model:                            OLS   Adj. R-squared:                  0.086
Method:                 Least Squares   F-statistic:                     1.584
Date:                Mon, 27 Nov 2023   Prob (F-statistic):              0.200
Time:                        01:37:20   Log-Likelihood:                -51.604
No. Observations:                  32   AIC:                             115.2
Df Residuals:                      26   BIC:                             124.0
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const          -0.4169      1.585     -0.263      

In [ ]:
#Takeaways

## Tackles in the defensive third: 

#### The coefficient for TklDef (Tackles in the defensive third) is -0.1108. Holding other variables constant, a one-unit increase in TklDef is associated with a decrease of 0.1108 units in goals conceded. 

#### Generally, this would indicate a good number of tackles in the defensive third is valuable the team. Liverpool in particular are a transition based team, so their ability to get the ball moving forward after making a tackle in the defensive third might be hinted by this statistic.

## Clearances:

#### The coefficient for Clr (clearances) is 0.0107. Holding other variables constant, a one-unit increase in Clr is associated with an increase of 0.0107 units in goals conceded. 

#### For a team that doesn’t like to clear the ball away and work on making their way to the front when they do win the ball, it makes sense that the higher the number of clearances, the more likely that they will concede. 




## A model to draw correlation between the touches in the different areas and chances

In [161]:
colsattack

['GF', 'GCA', 'SCA', 'TouchesDef', 'TouchesMid', 'TouchesAtt', 'TouchesPen']

In [162]:
df1 = cleaned_data[colsattack]

In [163]:
df1 = pd.DataFrame(df1)

In [164]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   GF          52 non-null     object 
 1   GCA         46 non-null     float64
 2   SCA         46 non-null     float64
 3   TouchesDef  46 non-null     float64
 4   TouchesMid  46 non-null     float64
 5   TouchesAtt  46 non-null     float64
 6   TouchesPen  46 non-null     float64
dtypes: float64(6), object(1)
memory usage: 3.0+ KB


In [166]:
df1.head(3)

GF  GCA   SCA  TouchesDef  TouchesMid  TouchesAtt  TouchesPen
0  3  NaN   NaN         NaN         NaN         NaN         NaN
1  2  4.0  20.0       260.0       349.0       179.0        33.0
2  1  2.0  45.0       147.0       527.0       244.0        42.0

In [170]:
df1 = df1.dropna()

In [172]:
df1[colsattack] = df1[colsattack].astype(int)

/var/folders/85/d_fv83j92kq22pp6q4hcdzmh0000gp/T/ipykernel_97578/3911405020.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1[colsattack] = df1[colsattack].astype(int)


In [173]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46 entries, 1 to 51
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   GF          46 non-null     int64
 1   GCA         46 non-null     int64
 2   SCA         46 non-null     int64
 3   TouchesDef  46 non-null     int64
 4   TouchesMid  46 non-null     int64
 5   TouchesAtt  46 non-null     int64
 6   TouchesPen  46 non-null     int64
dtypes: int64(7)
memory usage: 2.9 KB


In [178]:
y1 = df1["GCA"]

In [179]:
x1 = df1[['SCA', 'TouchesDef', 'TouchesMid', 'TouchesAtt', 'TouchesPen']]

In [180]:
model2 = sm.OLS(y1,x1).fit()

In [182]:
print(model2.summary())

                                 OLS Regression Results                                
Dep. Variable:                    GCA   R-squared (uncentered):                   0.646
Model:                            OLS   Adj. R-squared (uncentered):              0.603
Method:                 Least Squares   F-statistic:                              14.96
Date:                Mon, 27 Nov 2023   Prob (F-statistic):                    2.34e-08
Time:                        22:16:40   Log-Likelihood:                         -114.36
No. Observations:                  46   AIC:                                      238.7
Df Residuals:                      41   BIC:                                      247.9
Df Model:                           5                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

In [ ]:
#Takeaways

## Pinning Liverpool in their own half:

#### -> The coefficient for TouchesDef is -0.0112. A one-unit increase in TouchesDef is associated with a decrease of 0.0112 units in GCA (Goal Creating Actions)..

#### A fairly straight forward result as the more touches you make Liverpool take in their own half, the less likely they are to score, which is confirmed by the negative sign in TouchesDef.


## Avoiding transitions against Liverpool is key:

#### -> The coefficient for TouchesAtt is -0.0601. A one-unit increase in TouchesAtt is associated with a decrease of 0.0601 units in GCA.

#### This suggests the more touches Liverpool take in the attacking third, the less likely they are to create a goal creating action. This possibly points out their reliance on being a transition heavy team, and as a opposition team, perhaps the best way to not concede is to avoid as many transitions as possible and get men behind the ball.

## Defending in the box:

#### -> The coefficient for TouchesPen is 0.1768. A one-unit increase in TouchesPen is associated with an increase of 0.1768 units in GCA.

#### Inside the box, there’s a completely different story. He more touches Liverpool take in the box, the likelier it is to score. However, there Is likely to be a cutoff point that can be found using polynomial regression.


